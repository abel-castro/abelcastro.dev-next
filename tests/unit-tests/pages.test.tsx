import { render, screen } from '@testing-library/react';
import { Metadata } from 'next';
import React, { Suspense } from 'react';
import { afterEach, describe, expect, test, vi } from 'vitest';

import SinglePostPage, {
    SinglePostPageProps,
    generateMetadata,
} from '../../app/blog/[slug]/page';
import BlogHome from '../../app/blog/page';
import Home from '../../app/page';
import PrivacyPolicyPage from '../../app/privacy-policy/page';

describe('Home page', () => {
    test('Home page component should match the snapshot', async () => {
        const { container } = render(<Home />);

        expect(container).toMatchSnapshot();
    });
});
describe('Blog page', () => {
    afterEach(() => {
        vi.clearAllMocks();
        vi.resetAllMocks();
    });

    test('Blog page component should match the snapshot', async () => {
        const searchParams = {
            query: '',
            page: '1',
        };

        const { container } = render(
            <Suspense>
                <BlogHome searchParams={searchParams} />
            </Suspense>,
        );

        // it is necessary access to the screen first.
        // Otherwise, toMatchSnapshot will generate an empty snapshot
        await screen.findByText('Post 1');
        expect(container).toMatchSnapshot();
    });
});

describe('Single Post Page', () => {
    afterEach(() => {
        vi.clearAllMocks();
        vi.resetAllMocks();
    });

    test('Component should match the snapshot', async () => {
        const postSlug = 'post-1';
        const params = {
            slug: postSlug,
        };

        const { container } = render(
            <Suspense>
                <SinglePostPage params={params} />
            </Suspense>,
        );

        // it is necessary access to the screen first.
        // Otherwise, toMatchSnapshot will generate an empty snapshot
        await screen.findByText('Post 1');
        expect(container).toMatchSnapshot();
    });

    test('Component should raise not found', async () => {
        const postSlug = 'post-not-found';
        const params = {
            slug: postSlug,
        };

        const { container } = render(<SinglePostPage params={params} />);

        expect(container).toMatchSnapshot();
    });

    test('generateMetadata should return metadata for a valid post', async () => {
        const postSlug = 'post-1';

        const props: SinglePostPageProps = {
            params: { slug: postSlug },
        };

        const expectedMetadata: Metadata = {
            title: 'Post 1 | abelcastro.dev',
            description: 'Post 1 meta description',
        };

        const result = await generateMetadata(props);

        expect(result).toEqual(expectedMetadata);
    });

    test('generateMetadata should return null if the post is not found', async () => {
        const props: SinglePostPageProps = {
            params: { slug: 'non-existent-post' },
        };

        const result = await generateMetadata(props);

        expect(result).toBeNull();
    });
});

test('Privacy policy page component should match the snapshot', async () => {
    const { container } = render(<PrivacyPolicyPage />);

    expect(container).toMatchSnapshot();
});
