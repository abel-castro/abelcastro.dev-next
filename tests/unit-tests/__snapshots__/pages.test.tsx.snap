// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Blog page > Blog page component should match the snapshot 1`] = `
<div>
  <header
    class="flex flex-col justify-between sm:flex-row sm:justify-between p-4"
  >
    <h1
      class="text-3xl font-bold rainbow-text text-center sm:text-left content-center"
    >
      <a
        href=""
      >
        abelcastro.dev
      </a>
    </h1>
    <form
      class="relative flex mt-4 sm:mt-0 w-full sm:w-auto sm:flex justify-center"
      data-testid="search-form"
    >
      <label
        class="sr-only"
        for="search"
      >
        Search
      </label>
      <input
        class="peer block rounded border border-sky-500 py-[9px] lg:pl-8 pl-4 text-sm outline-2 dark:bg-slate-800 w-full max-w-xs"
        placeholder="Search in posts..."
      />
      <svg
        aria-hidden="true"
        class="hidden lg:block absolute left-3 top-1/2 h-[18px] w-[18px] -translate-y-1/2 text-gray-500 "
        data-slot="icon"
        fill="none"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m21 21-5.197-5.197m0 0A7.5 7.5 0 1 0 5.196 5.196a7.5 7.5 0 0 0 10.607 10.607Z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </form>
  </header>
  <main
    class="flex-grow p-4"
  >
    <div
      class="max-w-4xl mx-auto"
      data-testid="post-container"
      id="post-container"
    >
      <div
        class="post-element mb-8"
      >
        <h2
          class="text-2xl font-bold mb-2 text-sky-500  hover:text-sky-700"
        >
          <a
            href="blog/post-1"
          >
            Post 1
          </a>
        </h2>
        <p
          class="text-sm"
        >
          2024-08-16
        </p>
        <div
          class="flex space-x-2 mb-2"
        >
          <span
            class="text-xs bg-sky-100 text-gray-800 py-1 px-2 rounded"
          >
            Tag 1
          </span>
          <span
            class="text-xs bg-sky-100 text-gray-800 py-1 px-2 rounded"
          >
            Tag 2
          </span>
        </div>
        <div
          class="from-markdown max-w-none mt-4"
        >
          <div>
            Post 1 content
          </div>
        </div>
      </div>
      <hr
        class="my-12 h-px border-t-0 bg-transparent bg-gradient-to-r from-transparent via-sky-400 to-transparent opacity-50 dark:via-sky-300"
      />
      <div
        class="post-element mb-8"
      >
        <h2
          class="text-2xl font-bold mb-2 text-sky-500  hover:text-sky-700"
        >
          <a
            href="blog/post-2"
          >
            Post 2
          </a>
        </h2>
        <p
          class="text-sm"
        >
          2024-08-17
        </p>
        <div
          class="flex space-x-2 mb-2"
        >
          <span
            class="text-xs bg-sky-100 text-gray-800 py-1 px-2 rounded"
          >
            Tag 3
          </span>
          <span
            class="text-xs bg-sky-100 text-gray-800 py-1 px-2 rounded"
          >
            Tag 4
          </span>
        </div>
        <div
          class="from-markdown max-w-none mt-4"
        >
          <div>
            Post 2 content
          </div>
        </div>
      </div>
      <hr
        class="my-12 h-px border-t-0 bg-transparent bg-gradient-to-r from-transparent via-sky-400 to-transparent opacity-50 dark:via-sky-300"
      />
      <div
        class="mt-12 flex w-full justify-center"
      >
        <div
          class="inline-flex"
          data-testid="pagination"
        >
          <div
            class="flex h-10 w-10 items-center justify-center rounded-md border pointer-events-none text-white-700 mr-2 md:mr-4"
            data-testid="pagination-back"
          >
            <svg
              aria-hidden="true"
              class="w-4"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10.5 19.5 3 12m0 0 7.5-7.5M3 12h18"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </div>
          <div
            class="flex -space-x-px"
          >
            <div
              class="flex h-10 w-10 items-center justify-center text-sm border rounded-l-md rounded-r-md z-10 bg-teal-500 text-white-700"
            >
              1
            </div>
          </div>
          <div
            class="flex h-10 w-10 items-center justify-center rounded-md border pointer-events-none text-white-700 ml-2 md:ml-4"
            data-testid="pagination-forward"
          >
            <svg
              aria-hidden="true"
              class="w-4"
              data-slot="icon"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.5 4.5 21 12m0 0-7.5 7.5M21 12H3"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </main>
  <footer
    class="text-slate-300 p-4 text-center text-sm"
  >
    <p>
      Abel Castro 
      2025
       - checkout the source code of this page on
       
      <a
        class="footer-link link"
        href="https://github.com/abel-castro/abelcastro.dev-next"
        target="_blank"
      >
        GitHub
      </a>
       
      -
       
      <a
        class="footer-link link"
        href="/privacy-policy"
      >
        Privacy Policy
      </a>
       
    </p>
  </footer>
</div>
`;

exports[`Home page > Home page component should match the snapshot 1`] = `
<div>
  <header
    class="flex flex-col justify-between sm:flex-row sm:justify-between pl-4 pr-4"
  >
    <h1
      class="text-3xl font-bold rainbow-text text-center sm:text-left content-center"
    >
      <a
        href=""
      >
        abelcastro.dev
      </a>
    </h1>
    <div
      class="relative flex mt-4 sm:mt-0 w-full sm:w-auto sm:flex justify-center"
    >
      <ul
        class="menu menu-vertical sm:menu-horizontal font-bold text-lg link"
      >
        <li>
          <a
            href="#about-me"
          >
            About me
          </a>
        </li>
        <li>
          <a
            href="#blog"
          >
            My blog
          </a>
        </li>
        <li>
          <a
            href="#projects"
          >
            My projects
          </a>
        </li>
      </ul>
    </div>
  </header>
  <main
    class="lg:container mx-auto lg:max-w-8xl"
  >
    <div
      class="md:container md:max-w-5xl mx-auto hero min-h-screen"
      id="about-me"
    >
      <div
        class="hero-content flex-col md:flex-row-reverse"
      >
        <img
          alt="My younger self lifting bananas"
          class="w-[300px] sm:w-[400px] md:w-[500px] lg:w-[600px] xl:w-[800px] mask mask-hexagon shadow-2xl p-6"
          data-nimg="1"
          decoding="async"
          height="800"
          loading="lazy"
          src="/_next/image?url=%2Fpublic%2Fimages%2Fbanana-lifting.jpg&w=1920&q=75"
          srcset="/_next/image?url=%2Fpublic%2Fimages%2Fbanana-lifting.jpg&w=828&q=75 1x, /_next/image?url=%2Fpublic%2Fimages%2Fbanana-lifting.jpg&w=1920&q=75 2x"
          style="color: transparent;"
          width="800"
        />
        <div>
          <p
            class="text-xl"
          >
            Hi! Iâ€™m 
            <span
              class="font-bold"
            >
              Abel
            </span>
            , a
             
            <span
              class="font-bold"
            >
              Full Stack Developer
            </span>
             
            who loves coding and technology in general.
          </p>
          <p
            class="py-6"
          >
            That guy in the picture is me, back in 2008, before I began to code. I come from the beautiful Canary Island La Palma but for several years now, I have made Austria my home.
          </p>
          <p>
            Throughout most of my career, I was quite focused on developing with
             
            <span
              class="font-bold"
            >
              Python
            </span>
             and
             
            <span
              class="font-bold"
            >
              Django
            </span>
            , but in the last years, I have been learned to love
            <span
              class="font-bold"
            >
              TypeScript
            </span>
             
            and frameworks like
             
            <span
              class="font-bold"
            >
              Next.js
            </span>
             and
             
            <span
              class="font-bold"
            >
              NestJS
            </span>
            .
          </p>
        </div>
      </div>
    </div>
    <hr
      class="my-12 h-px border-t-0 bg-transparent bg-gradient-to-r from-transparent via-sky-400 to-transparent opacity-50 dark:via-sky-300"
    />
    <div
      class="md:container md:max-w-5xl mx-auto hero min-h-screen"
      id="blog"
    >
      <div
        class="hero-content"
      >
        <div>
          <h2
            class="text-3xl font-bold py-6"
          >
            My blog
          </h2>
          <p
            class="pb-3"
          >
            In my blog I (and LLMs ðŸ¤–) write about coding-related topics.
          </p>
          <p>
            The primary goal of my blog has always been to experiment and explore new technologies rather than getting more views.
          </p>
          <p>
            The initial version of my blog was launched in 2021 using Django, Django templates, and Bootstrap. Over time, the blog evolved significantly: it incorporated htmx and even implemented a REST API. Both the blog and this page were part of the same Django project.
          </p>
          <p
            class="pb-3 pt-3"
          >
            But similarly to my career path, the blog and the frameworks I work with have evolved significantly. When I discovered the charm of Next.js, I decided to re-implement it using Next.js. Check it out using the link below!
          </p>
          <p>
            Please note that all opinions expressed on the blog are my own.
          </p>
          <a
            class="btn text-white bg-sky-500 hover:bg-sky-700 mt-6"
            href="/blog"
          >
            To the blog ðŸš€
          </a>
        </div>
      </div>
    </div>
    <hr
      class="my-12 h-px border-t-0 bg-transparent bg-gradient-to-r from-transparent via-sky-400 to-transparent opacity-50 dark:via-sky-300"
    />
    <div
      class="md:container md:max-w-5xl mx-auto hero min-h-screen mb-10"
      id="projects"
    >
      <div
        class="hero-content"
      >
        <div>
          <h2
            class="text-3xl font-bold py-6"
          >
            My projects
          </h2>
          <div
            class="columns-3xs"
          >
            <div
              class="card mb-6 dark:bg-slate-800 shadow-xl w-full aspect-square undefined"
            >
              <figure>
                <img
                  alt="Fasting Timer 18:6"
                  data-nimg="1"
                  decoding="async"
                  height="250"
                  loading="lazy"
                  src="/_next/image?url=https%3A%2F%2Fgithub.com%2Fabel-castro%2Ffasting-timer%2Fblob%2Fmain%2Ffasting-timer-screenshot.jpg%3Fraw%3Dtrue&w=828&q=75"
                  srcset="/_next/image?url=https%3A%2F%2Fgithub.com%2Fabel-castro%2Ffasting-timer%2Fblob%2Fmain%2Ffasting-timer-screenshot.jpg%3Fraw%3Dtrue&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fgithub.com%2Fabel-castro%2Ffasting-timer%2Fblob%2Fmain%2Ffasting-timer-screenshot.jpg%3Fraw%3Dtrue&w=828&q=75 2x"
                  style="color: transparent;"
                  width="400"
                />
              </figure>
              <div
                class="card-body"
              >
                <h2
                  class="card-title"
                >
                  Fasting Timer 18:6
                </h2>
                <p>
                  Is a simple application designed to help users manage their intermittent fasting schedules.  I used React and Next.js to build it.
                </p>
                <div
                  class="mt-3 card-actions justify-end"
                >
                  <a
                    class="btn text-white bg-sky-500 hover:bg-sky-700 "
                    href="/blog/building-fasting-timer-for-learning-react-nextjs"
                  >
                    Read more
                  </a>
                </div>
              </div>
            </div>
            <div
              class="card mb-6 dark:bg-slate-800 shadow-xl w-full aspect-square undefined"
            >
              <figure>
                <img
                  alt="Sports Dashboard"
                  data-nimg="1"
                  decoding="async"
                  height="250"
                  loading="lazy"
                  src="/_next/image?url=https%3A%2F%2Fi.imgur.com%2FZchxi4B.png&w=828&q=75"
                  srcset="/_next/image?url=https%3A%2F%2Fi.imgur.com%2FZchxi4B.png&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fi.imgur.com%2FZchxi4B.png&w=828&q=75 2x"
                  style="color: transparent;"
                  width="400"
                />
              </figure>
              <div
                class="card-body"
              >
                <h2
                  class="card-title"
                >
                  Sports Dashboard
                </h2>
                <p>
                  Displays the results and standings of the most popular sports leagues. Made with an Angular frontend and a Django backend.
                </p>
                <div
                  class="mt-3 card-actions justify-end"
                >
                  <a
                    class="btn text-white bg-sky-500 hover:bg-sky-700 "
                    href="/blog/sports-dashboard-displays-last-matchday-results"
                  >
                    Read more
                  </a>
                </div>
              </div>
            </div>
            <div
              class="card mb-6 dark:bg-slate-800 shadow-xl w-full aspect-square undefined"
            >
              <figure>
                <img
                  alt="Basic analytics"
                  data-nimg="1"
                  decoding="async"
                  height="250"
                  loading="lazy"
                  src="/_next/image?url=https%3A%2F%2Fi.imgur.com%2FvqKIa0M.png&w=828&q=75"
                  srcset="/_next/image?url=https%3A%2F%2Fi.imgur.com%2FvqKIa0M.png&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fi.imgur.com%2FvqKIa0M.png&w=828&q=75 2x"
                  style="color: transparent;"
                  width="400"
                />
              </figure>
              <div
                class="card-body"
              >
                <h2
                  class="card-title"
                >
                  Basic analytics
                </h2>
                <p>
                   My self made and hosted alternative to Google Analytics. Made with Django, DRF, Bootstrap and Chart.js.
                </p>
                <div
                  class="mt-3 card-actions justify-end"
                >
                  <a
                    class="btn text-white bg-sky-500 hover:bg-sky-700 "
                    href="/blog/my-self-made-hosted-google-analytics-alternative"
                  >
                    Read more
                  </a>
                </div>
              </div>
            </div>
            <div
              class="card mb-6 dark:bg-slate-800 shadow-xl w-full aspect-square undefined"
            >
              <figure>
                <img
                  alt="Kudos-Box"
                  data-nimg="1"
                  decoding="async"
                  height="250"
                  loading="lazy"
                  src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2Fabel-castro%2Fkudos-box%2Fmain%2Fscreenshot.jpg&w=828&q=75"
                  srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2Fabel-castro%2Fkudos-box%2Fmain%2Fscreenshot.jpg&w=640&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2Fabel-castro%2Fkudos-box%2Fmain%2Fscreenshot.jpg&w=828&q=75 2x"
                  style="color: transparent;"
                  width="400"
                />
              </figure>
              <div
                class="card-body"
              >
                <h2
                  class="card-title"
                >
                  Kudos-Box
                </h2>
                <p>
                  Anonymously send kudos to your colleagues and friends. The project is not online anymore but the source code is available on GitHub.
                </p>
                <div
                  class="mt-3 card-actions justify-end"
                >
                  <a
                    class="btn text-white bg-sky-500 hover:bg-sky-700 "
                    href="https://github.com/abel-castro/kudos-box/"
                  >
                    Read more
                  </a>
                </div>
              </div>
            </div>
            <div
              class="card mb-6 dark:bg-slate-800 shadow-xl w-full aspect-square undefined"
            >
              <div
                class="card-body"
              >
                <h2
                  class="card-title"
                >
                  Other Stuff
                </h2>
                <p>
                  I regularly work on other projects and experiment with new ideas. Check out my GitHub profile to see what Iâ€™ve been up to.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer
      class="text-slate-300 p-4 text-center text-sm"
    >
      <p>
        Abel Castro 
        2025
         - checkout the source code of this page on
         
        <a
          class="footer-link link"
          href="https://github.com/abel-castro/abelcastro.dev-next"
          target="_blank"
        >
          GitHub
        </a>
         
        -
         
        <a
          class="footer-link link"
          href="/privacy-policy"
        >
          Privacy Policy
        </a>
         
      </p>
    </footer>
  </main>
</div>
`;

exports[`Privacy policy page component should match the snapshot 1`] = `
<div>
  <main>
    <h2
      class="text-2xl font-bold mb-8"
    >
      Privacy Policy for abelcastro.dev
    </h2>
    <p>
      This Privacy Policy document outlines the types of information that is collected and recorded by blog.abelcastro.dev and how I use it.
    </p>
    <h3
      class="font-bold mt-4"
    >
      1. Information We Collect
    </h3>
    <p>
      This service collects data regarding visitors request metadata, which includes IP addresses, browser types, operating systems, and other similar information. This data is solely used for statistical analysis purposes to improve our website and services. No personal data is intentionally collected or used for profiling or marketing purposes.
    </p>
    <h3
      class="font-bold mt-4"
    >
      2. Use of Collected Information
    </h3>
    <p>
      The information I collect is used exclusively for generating aggregate statistical insights. These insights help me understand how visitors interact with the website, enabling me to enhance user experience and optimize the content. The data is not shared with or sold to third parties, including external analytics platforms like Google Analytics.
    </p>
    <h3
      class="font-bold mt-4"
    >
      3. Data Processing and Storage
    </h3>
    <p>
      The collected data is processed and stored securely. I take appropriate measures to prevent unauthorized access, disclosure, modification, or unauthorized destruction of the data.
    </p>
    <h3
      class="font-bold mt-4"
    >
      4. Open Source Analytics
    </h3>
    <p>
      In the spirit of transparency, my analytics tool is open-source, and the source code can be reviewed at
       
      <a
        href="https://github.com/abel-castro/basic_analytics"
      >
        https://github.com/abel-castro/basic_analytics prefetch=
      </a>
      .
    </p>
    <h3
      class="font-bold mt-4"
    >
      5. Your Privacy Rights
    </h3>
    <p>
      As a visitor, you have the right to understand how your data is being used. Should you have any questions or concerns about I use of your request metadata, please feel free to contact me.
    </p>
    <h3
      class="font-bold mt-4"
    >
      6. Consent
    </h3>
    <p>
      By using this website, you hereby consent to our Privacy Policy and agree to its terms.
    </p>
    <h3
      class="font-bold mt-4"
    >
      7. Policy Updates
    </h3>
    <p>
      I reserve the right to update this policy as necessary to stay compliant with relevant laws and to reflect any changes to our data collection and usage practices. Any updates will be posted on this page.
    </p>
    <h3
      class="font-bold mt-4"
    >
      8. Hosting and Analytics Tools
    </h3>
    <p>
      This website is hosted on the Vercel platform. I use in this website Vercel
       
      <a
        href="https://vercel.com/docs/analytics"
      >
        Analytics
      </a>
       
      and
       
      <a
        href="https://vercel.com/docs/speed-insights"
      >
        Speed Insights
      </a>
       
      tools to monitor and optimize the performance of this website. These tools collect data on site performance, such as load times and error rates, to help improve the overall user experience. Vercels use of this data is governed by their own privacy policies, which can be reviewed on their website.
    </p>
    <p>
      While Vercel may collect information related to site performance, such as server logs and performance metrics, this data is used solely to enhance the performance and reliability of the website. No personally identifiable information is shared with Vercel beyond what is necessary for the operation of their services.
    </p>
  </main>
</div>
`;

exports[`Single Post Page > Component should match the snapshot 1`] = `
<div>
  <main>
    <div
      class="post-element mb-8"
    >
      <h2
        class="text-2xl font-bold mb-2 text-sky-500  "
      >
        Post 1
      </h2>
      <p
        class="text-sm"
      >
        2024-08-16
      </p>
      <div
        class="flex space-x-2 mb-2"
      >
        <span
          class="text-xs bg-sky-100 text-gray-800 py-1 px-2 rounded"
        >
          Tag 1
        </span>
        <span
          class="text-xs bg-sky-100 text-gray-800 py-1 px-2 rounded"
        >
          Tag 2
        </span>
      </div>
      <div
        class="from-markdown max-w-none mt-4"
      >
        <div>
          Post 1 content
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`Single Post Page > Component should raise not found 1`] = `<div />`;
